File: 算法与数据结构/python实现数据结构/stack_/stack.py
# coding:utf-8
from array_.array import Array


class ArrayStack(object):
    """
    基于动态数组基础上实现的栈
    """
    def __init__(self, capacity):
        self.array = Array(capacity)

    def push(self, e):
        """
        想栈中加入新元素
        :return:
        """
        return self.array.add_last(e)

    def pop(self):
        """
        从栈中取出栈顶元素
        :return:
        """
        return self.array.remove_last()

    def peek(self):
        """
        查询栈顶元素个数
        :return:
        """
        return self.array.get_last()

    def get_size(self):
        """
        查询栈元素个数
        :return:
        """
        return self.array.get_size()

    def get_capacity(self):
        """
        查询栈容量
        :return:
        """
        return self.array.get_capacity()

    def is_empty(self):
        """
        判断栈是否为空
        :return:
        """
        return self.array.is_empty()

    def __str__(self):
        res = "Stack: ["
        for i in range(self.array.get_size()):
            res += "{0}".format(self.array.get(i))
            if i != self.array.get_size() - 1:
                res += ", "
        res += "] top"

        return res


def test_array_stack():
    st = ArrayStack(10)
    for i in ["test1", "test2", "test3"]:
        st.push(i)
    print(st)
    st.pop()
    print(st)
    st.pop()
    print(st)
    st.pop()
    print(st)


if __name__ == "__main__":
    test_array_stack()

File: 算法与数据结构/golang实现数据结构/DataStructAndAlgorithm/stack/stack/stack.go
package stack

import (
	"DataStructAndAlgorithm/stack/array"
	"fmt"
)

//在已实现的动态数组基础上通过接口实现栈
 type ArrayStack struct {
	 array *array.Array
 }

// 新建ArrayStack对象实例，指定容量
func NewArrayStack(capacity int) *ArrayStack {
	arr := array.NewArray(capacity)
	stack := ArrayStack{array: arr}
	return &stack
}

// 新建ArrayStack对象实例，使用默认容量
func NewArrayStack1() *ArrayStack {
	arr := array.NewArray1()
	stack := ArrayStack{array: arr}
	return &stack
}

// 向栈中加入新元素
func (st *ArrayStack) Push(e interface{}) error {
	return st.array.AddLast(e)
}

// 从栈中取出栈顶元素
func (st *ArrayStack) Pop() (interface{}, error) {
	return st.array.RemoveLast()
}

// 查询栈顶元素
func (st *ArrayStack) Peek() (interface{}, error){
	return st.array.GetLast()
}

// 查询栈元素个数
func (st *ArrayStack) GetSize() int {
	return st.array.GetSize()
}

// 查询栈的容量
func (st *ArrayStack) GetCapacity() int {
	return st.array.GetCapacity()
}

// 判断栈是否为空
func (st *ArrayStack) IsEmpty() bool {
	return st.array.IsEmpty()
}

// 格式化输出栈信息
func (st *ArrayStack) String() string{
	res := fmt.Sprintf("Stack: [")
	for i:=0; i < st.array.GetSize(); i++ {
		data, _ := st.array.Get(i)
		res += fmt.Sprintf("%v", data)
		if i != st.array.GetSize() - 1{
			res += ", "
		}
	}
	res += "] top"
	return  res
}
